### ISSUE TITLE
Cannot set aria-disabled on Button component

### ISSUE THREAD

**Issue by kylemh** [adobe/react-spectrum#8687]
### Provide a general summary of the issue here

I'd like to use `aria-disabled` instead of `isDisabled` in some situations.

### 🤔 Expected Behavior?

Defining `aria-disabled` as `true` or `false` should apply the attribute to the element in the DOM

### 😯 Current Behavior

Defining aria-disabled on a Button yields no type error; however, no value is ever applied to the actual element in the DOM.

### 🖥️ Steps to Reproduce

Render a Button component from react-aria-components, and apply an `aria-disabled` prop with `true` or `false` as the value. Look for the relevant element in devtools to notice that the attribute isn't defined.

### Version

1.11.0

### What browsers are you seeing the problem on?

Chrome

### What operating system are you using?

MacOS

**Comment by snowystinger** [comment]
> I'd like to use aria-disabled instead of isDisabled in some situations.

To discuss this, it'd be helpful to know more about those situations. Do you have some real life examples?

**Comment by kylemh** [comment]
I think [this article](https://adamsilver.io/blog/the-problem-with-disabled-buttons-and-what-to-do-instead/) represents many good points. Another very good point is that sometimes I'd like to provide seeing-eye users clues as to why a button is disabled via Tooltip, but I can't do that when the button is **actually** disabled because Tooltips dont render on non-focusable elements.

The reality is, sometimes there are situations where a disabled button is valid and sometimes there are situations where an aria-disabled button is valid.

In one of the internal apps that I work on with users of assistive technology, we actually defaulted to only doing `aria-disabled` and prevent the click handler if that attribute is true. In doing so, we feel we're providing the optimal experience users who do and do not leverage assistive technology.

**Comment by snowystinger** [comment]
I agree, there are cases for using `aria-disabled`. For example, we use it for Pending button state. https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/Button.tsx#L150

It'd be nice to determine if there are some other patterns that this should apply to, that way we can help users know when it's appropriate or not. I think if anything that article highlights the difficulty in knowing what you should be doing.

I think that setting aria-disabled only can be a false flag. It's *only* telling AT users that the button is disabled. You'd need to communicate in a different way that the button is disabled to sighted users, and if that's a tooltip, then you'd need to communicate it in *yet another* way to touch users.

As for that article, I think leaving the button enabled is the way to go. Validation and focus moving to the first invalid field would explain why the action wasn't taken. Also, the article doesn't actually make a recommendation, it hides that behind a paywall, and it never mentions aria-disabled.

All of that said, I'm not strictly against adding support for aria-disabled. We'd need to determine if only having it on button was correct and if it should automatically disable any interactions (probably shouldn't but then users have to remember to implement that themselves).

**Comment by kylemh** [comment]
> I think that setting aria-disabled only can be a false flag. It's only telling AT users that the button is disabled. You'd need to communicate in a different way that the button is disabled to sighted users, and if that's a tooltip, then you'd need to communicate it in yet another way to touch users.

We add CSS styling conditionally based on the presence of `aria-disabled="true"` much the same way one would do via `:disabled`. The tooltip is optional and useful in some situations.

Can you think of anything wrong with this pattern when a react-aria-components Button has `isDisabled={true}`?

- Never set a button's `disabled` to `true`
- `aria-disabled` and `data-disabled` set to `true`
- Callbacks for `onPress`/`onClick` are prevented

---

I am very confused by https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/Button.tsx#L150 - my OP is that it's not possible to define the value for `aria-disabled` with RAC Button... this code makes it look like it SHOULD be possible.

**Comment by snowystinger** [comment]
> I am very confused by https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/Button.tsx?rgh-link-date=2025-08-08T00%3A01%3A26.000Z#L150 - my OP is that it's not possible to define the value for `aria-disabled` with RAC Button... this code makes it look like it SHOULD be possible.

it gets stripped out here https://github.com/adobe/react-spectrum/blob/155e361c6e2910c0b087ccba7cc61541fb50b0a7/packages/%40react-aria/button/src/useButton.ts#L108

> Can you think of anything wrong with this pattern when a react-aria-components Button has isDisabled={true}?

I believe we considered this at one point, @majornista do you recall?

One issue I can think of is that implicit form submission would probably still work.

I think it would also be bad for keyboard users to focus something that they can't really see, since disabled controls don't need to meet contrast.

**Comment by kylemh** [comment]
> I think it would also be bad for keyboard users to focus something that they can't really see, since disabled controls don't need to meet contrast.

For this point specifically, my opinion would be that it's probably easy to dismiss because keyboard navigating users tab through plenty of elements that don't look focusable despite being focusable.

> One issue I can think of is that implicit form submission would probably still work.

As far as I can tell, if the submit button for the form uses the same logic I laid out earlier, then no the form will not submit because its click handler has `preventDefault()` when `aria-disabled` is true. This is true when the submit button is inside the form and if it's connected via `form` prop.

**Comment by snowystinger** [comment]
Talked about this with the team today. We are ok to allow aria-disabled through to the dom, but we don't want to change browser behaviour, so we do not want to disable the press/click/hover/etc callbacks as a result of applying this.

**Comment by kylemh** [comment]
That seems totally fine to me and was my original request anyways! It can be up to me in my own apps/libraries to enact that click disable behavior.

**Comment by snowystinger** [comment]
Accidental merge, I have reverted so we can review it

**Comment by kylemh** [comment]
Hey @snowystinger sorry to revive this one, but I can't actually do the behavior I was hoping for...

```tsx
onClick={(e) => {
  if (isDisabled || isLoading) {
    e.preventDefault();
    e.stopPropagation();
    return;
  }
  onClick?.(e);
}}
```

When `isDisabled` is true, the press/click event still fires.

I can try to create a reproduction for you if you'd like, but I'm curious if my mistake is obvious? I know this would be a new issue, but I felt like this thread is a good place to chat first since it provides all the context for why I'm doing something seemingly egregious 😂
